### PR TITLE
Fix PG upgrade scripts for 9.4, 9.5, 10.0, 10.1

### LINKED ISSUE #5068: Duplicate key violation pg_dist_object_pkey during citus_finish_pg_upgrade
We've observed primary key violations during the citus_finish_pg_upgrade step:
```sql
ERROR: duplicate key value violates unique constraint "pg_dist_object_pkey"
DETAIL:  Key (classid, objid, objsubid)=(3079, 16425, 0) already exists.                                                                                                           
CONTEXT:  SQL statement "WITH old_records AS (                                                                                                                                     
        DELETE FROM                                                                                                                                                                
            citus.pg_dist_object                                                                                                                                                   
        RETURNING                                                                                                                                                                  
            type,                                                                                                                                                                  
            object_names,                                                                                                                                                          
            object_args,                                                                                                                                                           
            distribution_argument_index,                                                                                                                                           
            colocationid                                                                                                                                                           
    )                                                                                                                                                                              
    INSERT INTO citus.pg_dist_object (classid, objid, objsubid, distribution_argument_index, colocationid)                                                                         
    SELECT                                                                                                                                                                         
        address.classid,                                                                                                                                                           
        address.objid,                                                                                                                                                             
        address.objsubid,                                                                                                                                                          
        naming.distribution_argument_index,                                                                                                                                        
        naming.colocationid                                                                                                                                                        
    FROM                                                                                                                                                                           
        old_records naming,                                                                                                                                                        
        pg_get_object_address(naming.type, naming.object_names, naming.object_args) address
```
There were no actual duplicates in the table.

This seems to be general PG behaviour:
```sql
CREATE TABLE test(id int primary key);
INSERT INTO test SELECT i FROM generate_series(1,10) i;
WITH old_records AS (
DELETE FROM
test
RETURNING
id
)
INSERT INTO test (id)
SELECT
old_records.id + 1
FROM
old_records;
ERROR: duplicate key value violates unique constraint "test_pkey"
DETAIL: Key (id)=(2) already exists.
```

### PR DESCRIPTION
**Pull request by marcocitus**
DESCRIPTION: Fixes an issue that could cause citus_finish_pg_upgrade to fail

Rewiring Citus upgrade scripts to fix the prepare/finish PG upgrade scripts. 

Users who upgrade to the patch release of 9.4 will get the fix via 9.4-1--9.4-2.
Users who upgrade to the patch release of 9.5 will get the fix via 9.5-1--9.5-2.
Users who upgrade to the patch release of 10.0 will get the fix via 10.0-3--10.0-4.
Users who upgrade to Citus 10.1 will also get it as part of the 10.0-3--10.0-4 upgrade path.

Given that we use CREATE OR REPLACE, it's ok to get the fix multiple times.

Fixes #5068, but not #5069.

TODO:
- [x] Tests
- [x] Downgrades? (probably not)
- [ ] ~Older versions?~

Backports:
 - Citus 9.4 #5081
 - Citus 9.5 #5083
 - Citus 10.0 #5084

Verify upgrades locally
 - [x] 94 -> 9.5
 - [x] 9.4 -> 10.0
 - [x] 9.4 -> 10.1
 - [x] 9.5 -> 10.0
 - [x] 9.5 -> 10.1
 - [x] 10.0 -> 10.1

### CODE REVIEW

**Comment by thanodnl** [comment]
Solution looks solid to me

I think we saw this in production on extensions colliding on their object id right? Do we have a reliable way of hitting that? Such a reliable reproduction will be a good way for adding tests. If we don't have a reliable way I think we can't easily add automated tests for this issue.

**Comment by marcocitus** [comment]
When cherry-picking, keep in mind that:
- We should include commits for all preceding versions to ensure upgrade paths from patched version to patched version
- We need to generate a new latest.sql (check-sql-snapshots will probably fail)

While reviewing, keep in mind that GitHub hides UDF files with a specific SQL versions (e.g. 9.5-2.sql), but it has meaningful changes compared to the previous version of the function.

**Comment by thanodnl** [comment]
> GitHub hides UDF files with a specific SQL versions (e.g. 9.5-2.sql), but it has meaningful changes compared to the previous version of the function.

They should _not_ contain changes iirc. They should be new files that are introduced for the version we are introducing, equal to what latest should be in the specific version. I will need to understand how it is that they contain changes.

**Comment by marcocitus** [comment]
> They should be new files that are introduced for the version we are introducing, equal to what latest should be in the specific version. I will need to understand how it is that they contain changes.

I meant changes compared to the preceding version of the function (e.g. `diff 9.2-1.sql 9.4-2.sql`).

**Comment by thanodnl** [comment]
ah, yeah, makes sense.

Those changes will be visible on successful cherrypick to those releases as our CI enforces latest.sql to be equal to the new files introduced. We will seem the meaningful changes there.

One thing that I realise while reviewing commit per commit.
We will need to update the `default_version` in the `citus.control` file for every backported release we are making.

**Comment by thanodnl** [comment]
I think we _should_ add downgrade scripts as well.

Without downgrade scripts we break the downgrade path.
Including a downgrade, either no-op or factual downgrade, we don't break the path.

Postgres will take the shortest path anyway, meaning it will not deviate from the normal upgrade path if an unpatched version is upgraded to a later version, either patched or unpatched.

Lastly adding the downgrade path is relatively easy. Allowing us to retain the downgrade possibility from these versions.

Ideally we would rely _only_ on the downgrade path, as it makes the subsequent updates easier. Unfortunately we don't have the downgrade packages released. Having this fix blocked till we release these packages would delay the fix undesirably.

**Review by SaitTalhaNisanci — APPROVED** [review]

File: src/test/regress/expected/multi_extension.out [hunk 1]
```diff
@@ -388,6 +388,28 @@ SELECT * FROM print_extension_changes();
                  | function worker_save_query_explain_analyze(text,jsonb) SETOF record
 (2 rows)
 
+-- Test upgrade paths for backported citus_pg_upgrade functions
+ALTER EXTENSION citus UPDATE TO '9.4-2';
+ALTER EXTENSION citus UPDATE TO '9.4-1';
+-- Should be empty result since upgrade+downgrade should be a no-op
```
Comment (JelteF) [inline src/test/regress/expected/multi_extension.out:394]: This comment is a lie. It's not a no-op in this case, since we explicitly keep the new version during the downgrade. The result from print_extension_changes is still empty, but that's simply because arguments don't change.

Comment (thanodnl) [reply src/test/regress/expected/multi_extension.out:394]: makes sense, will address this subtlety in the comments

**Comment by thanodnl** [comment]
For older versions:

There could be reasons to support 8.3 as it is the last version in the previous major.
However, to backport this to 8.3 we will need to cover every version in between.
This would result in 3-4 more branches and 26 more manual tests to verify all upgrades work.

Our managed services have a runbook on how to overcome the upgrade problems when customers on older version run into this issue.

TLDR; copy contents of `pg_dist_object` to a backup table, truncate before calling the finish pg_upgrade function, insert into old catalog table from backup table with rewrite of the oid's.

It would have been significantly easier to _only_ backport to 8.3 if we had the downgrade packages distributed. We could make the 8.3 backport rely on 10.1+ for the downgrade packages, bringing back all installations to the main upgrade path. However we don't have them. If in a future we have, we could backport the patch to 8.3 somewhat easily.
